sdks/python/apache_beam/io/gcp/bigtableio_it_test.py

```diff
@@ -0,0 +1,178 @@
+
+  def _generate(self):
+    for i in range(ROW_COUNT):
+      key = "key_%s" % ('{0:012}'.format(i))
+      test_row = row.DirectRow(row_key=key)
+      value = ''.join(random.choice(LETTERS_AND_DIGITS) for _ in range(CELL_SIZE))
+      for j in range(COLUMN_COUNT):
+        test_row.set_cell(column_family_id=COLUMN_FAMILY_ID,
+                          column=('field%s' % j).encode('utf-8'),
+                          value=value,
+                          timestamp=datetime.datetime.now())
+      yield test_row
+
+  def expand(self, pvalue):
+    return (pvalue
+            | beam.Create(self._generate())
+            | WriteToBigTable(project_id=self.beam_options['project_id'],
+                              instance_id=self.beam_options['instance_id'],
+                              table_id=self.beam_options['table_id']))
+
+@unittest.skipIf(Client is None, 'GCP Bigtable dependencies are not installed')
+class BigtableIOTest(unittest.TestCase):
+  """ Bigtable IO Connector Test
+
+  This tests the connector both ways, first writing rows to a new table, then reading them and comparing the counters
+  """
+  def setUp(self):
+    self.result = None
+    self.table = Client(project=PROJECT_ID, admin=True)\
+                    .instance(instance_id=INSTANCE_ID)\
+                    .table(TABLE_ID)
+
+    if not self.table.exists():
+      column_families = {COLUMN_FAMILY_ID: column_family.MaxVersionsGCRule(2)}
+      self.table.create(column_families=column_families)
+      logging.info('Table {} has been created!'.format(TABLE_ID))
+
+  def test_bigtable_io(self):
+    print('Project ID: {}'.format(PROJECT_ID))
+    print('Instance ID:{}'.format(INSTANCE_ID))
+    print('Table ID:   {}'.format(TABLE_ID))
+
+    pipeline_options = PipelineOptions(PIPELINE_PARAMETERS)
+    p = beam.Pipeline(options=pipeline_options)
+    _ = (p | 'Write Test Rows' >> GenerateTestRows())
+
+    self.result = p.run()
+    self.result.wait_until_finish()
+
+    assert self.result.state == PipelineState.DONE
+
+    if not hasattr(self.result, 'has_job') or self.result.has_job:
+      query_result = self.result.metrics().query(MetricsFilter().with_name('Written Row'))
+      if query_result['counters']:
+        read_counter = query_result['counters'][0]
+        logging.info('Number of Rows written: %d', read_counter.committed)
+        assert read_counter.committed == ROW_COUNT
+
+    p = beam.Pipeline(options=pipeline_options)
+    count = (p
+             | 'Read from Bigtable' >> Read(BigtableSource(project_id=PROJECT_ID,
+                                                           instance_id=INSTANCE_ID,
+                                                           table_id=TABLE_ID))
+             | 'Count Rows' >> Count.Globally())
+    self.result = p.run()
+    self.result.wait_until_finish()
+    assert_that(count, equal_to([ROW_COUNT]))
+
+
+if __name__ == '__main__':
+  parser = argparse.ArgumentParser()
```

Comment:
Please manually run this test several times and make sure it passes consistently ? Also approximately how long does it take for this test to run ?

Reply:
The test has been run successfully many times. It takes around 10-15 minutes for Dataflow to complete both, read and write parts. It will be repeated after making the necessary changes.